src/seat/pointer.rs

```diff
@@ -7,124 +7,87 @@ use wayland_client::{
 
-/// Describes a kind of pointer axis event.
-#[derive(Debug, Clone, Copy)]
-pub enum AxisKind {
-    /// The axis scrolling is in an absolute number of pixels.
-    Absolute(f64),
+/// Describes a scroll along one axis
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+pub struct AxisScroll {
+    /// The scroll measured in pixels.
+    pub absolute: f64,
 
-    /// The axis scrolling is in discrete units of lines or columns.
-    Discrete(i32),
-
-    /// The axis scrolling was stopped.
+    /// The scroll measured in steps.
     ///
-    /// Generally this variant is encountered when hardware indicates the end of some continuous scrolling.
-    Stop,
-}
+    /// Note: this might always be zero if the scrolling is due to a touchpad or other continuous
+    /// source.
+    pub discrete: i32,
 
-/// A type representing a scroll event.
-///
-/// A scroll event may consist of a vertical and horizontal component.
-#[derive(Debug)]
-pub struct PointerScroll {
-    horizontal: Option<AxisKind>,
-    vertical: Option<AxisKind>,
-    source: Option<wl_pointer::AxisSource>,
+    /// The scroll was stopped.
+    ///
+    /// Generally this is encountered when hardware indicates the end of some continuous scrolling.
+    pub stop: bool,
 }
 
-impl PointerScroll {
-    pub fn axis(&self, axis: wl_pointer::Axis) -> Option<AxisKind> {
-        match axis {
-            wl_pointer::Axis::VerticalScroll => self.vertical,
-            wl_pointer::Axis::HorizontalScroll => self.horizontal,
-
-            _ => unreachable!(),
-        }
+impl AxisScroll {
+    /// Returns true if there was no movement along this axis.
+    pub fn is_none(&self) -> bool {
+        *self == Self::default()
     }
 
-    pub fn source(&self) -> Option<wl_pointer::AxisSource> {
-        self.source
+    fn merge(&mut self, other: &Self) {
+        self.absolute += other.absolute;
+        self.discrete += other.discrete;
+        self.stop |= other.stop;
     }
+}
 
-    pub fn has_axis(&self, axis: wl_pointer::Axis) -> bool {
-        match axis {
-            wl_pointer::Axis::VerticalScroll => self.vertical.is_some(),
-            wl_pointer::Axis::HorizontalScroll => self.horizontal.is_some(),
-
-            _ => unreachable!(),
-        }
-    }
+/// A single pointer event.
+#[derive(Debug, Clone)]
+pub struct PointerEvent {
+    pub surface: wl_surface::WlSurface,
+    pub position: (f64, f64),
+    pub kind: PointerEventKind,
 }
 
-pub trait PointerHandler: SeatHandler + Sized {
-    /// The pointer focus is set to a surface.
-    ///
-    /// The `entered` parameter are the surface local coordinates from the top left corner where the cursor
-    /// has entered.
-    ///
-    /// The pos
-    fn pointer_focus(
-        &mut self,
-        conn: &Connection,
-        qh: &QueueHandle<Self>,
-        pointer: &wl_pointer::WlPointer,
-        surface: &wl_surface::WlSurface,
-        entered: (f64, f64),
+#[derive(Debug, Clone)]
+pub enum PointerEventKind {
```

Comment:
Could this be copy?

Reason being that the pointer handler gives you a slice of events, which you need to clone or pass a vec to the user?

Maybe even a tinyvec since often there are very few events per frame (4 before leaking to the heap I guess)?

Reply:
Plus the PointerEvent is trivial to copy.

Reply:
Using a smallvec for the events would make sense.  Making the kind `Copy` is possible, but `PointerEvent` can't be, so you lose most of the help that can provide.

Reply:
oh yeah the surface...

Reply:
I'd prefer tinyvec (no unsafe code) but smallvec is fine.

Reply:
I just chose `smallvec` because it's already a dependency.

Reply:
We can migrate wayland-rs to tinyvec if it's a better alternative btw, I haven't looked into this yet.

Reply:
After looking at the API, we can't use tinyvec here because there is no `impl Default for WlSurface`.

Reply:
> We can migrate wayland-rs to tinyvec if it's a better alternative btw, I haven't looked into this yet.

Yeah with the Default requirement in tinyvec it won't work here at least.

I believe you do end up with a slightly bigger struct in tinyvec because smallvec can use a union to save some space.